src/utils/message-for.ts

```diff
@@ -63,3 +93,114 @@ export function messageFor(
 
   return null;
 }
+
+export function messageForStorage(world: HubsWorld, created: EntityID[], updated: EntityID[], deleted: EntityID[]) {
+  const message: StorableMessage = {
+    version: 1,
+    creates: [],
+    updates: [],
+    deletes: []
+  };
+
+  created.forEach(eid => {
+    const { prefabName, initialData } = createMessageDatas.get(eid)!;
+    message.creates.push([APP.getString(Networked.id[eid])!, prefabName, initialData]);
+  });
+
+  updated.forEach(eid => {
+    const updateMessage: StorableUpdateMessage = {
+      nid: APP.getString(Networked.id[eid])!,
+      lastOwnerTime: Networked.lastOwnerTime[eid],
+      timestamp: Networked.timestamp[eid],
+      owner: APP.getString(Networked.owner[eid])!,
+      creator: APP.getString(Networked.creator[eid])!,
+      data: {}
+    };
+
+    for (let j = 0; j < networkableComponents.length; j++) {
+      const component = networkableComponents[j];
+      if (hasComponent(world, component, eid)) {
+        const schema = schemas.get(component)!;
+        updateMessage.data[schema.componentName] = schema.serializeForStorage(eid);
+      }
+    }
+
+    message.updates.push(updateMessage);
+  });
+
+  deleted.forEach(eid => {
+    // TODO: We are reading component data of a deleted entity here.
+    const nid = Networked.id[eid];
+    message.deletes.push(APP.getString(nid)!);
+  });
+
+  if (message.creates.length || message.updates.length || message.deletes.length) {
+    return message;
+  }
+
+  return null;
+}
+
+export interface LegacyRoomObject {
+  extensions: {
+    HUBS_components: {
+      media: {
+        version: number;
+        src: string;
+        id: NetworkID;
+        contentSubtype?: string;
+      };
+      pinnable: {
+        pinned: boolean;
+      };
+    };
+  };
+  name: NetworkID;
+  rotation: [number, number, number, number];
+  translation: [number, number, number];
+}
+
+export function messageForLegacyRoomObjects(objects: LegacyRoomObject[]) {
+  const message: Message = {
+    creates: [],
+    updates: [],
+    deletes: []
+  };
+
+  objects.forEach(obj => {
+    const nid = obj.name;
+    const initialData: MediaLoaderParams = {
+      src: obj.extensions.HUBS_components.media.src,
+      resize: false,
+      recenter: false,
+      animateLoad: false,
+      isObjectMenuTarget: true
+    };
+    const createMessage: CreateMessage = [nid, "media", initialData];
+    message.creates.push(createMessage);
+
+    const updateMessage: StorableUpdateMessage = {
+      data: {
+        "networked-transform": {
+          version: 1,
+          data: {
+            position: obj.translation,
+            rotation: obj.rotation,
+            scale: [1, 1, 1]
```

Comment:
We should also read scale, and also note that any of these can be undefined. https://github.com/mozilla/hubs/blob/master/src/utils/pinned-entity-to-gltf.js#L20-L22